### ISSUE TITLE
Check early stopping during pre-train routine

### ISSUE THREAD

**Issue by collinmccarthy** [PyTorchLightning/pytorch-lightning#8278]
## 🚀 Feature
Same as the [discussion here](https://github.com/PyTorchLightning/pytorch-lightning/discussions/6740), I would like the trainer to check the state of the early stopping callback during the pre-train routine, and signal the trainer to stop training before the first training epoch if it should stop according to the early stopping callback state.

### Motivation

When training stops due to early stopping, but testing crashes, the run needs to be resumed. If I let the system simply resume normally it will always run one training epoch, which could lead to many more epochs if the early stopping wait count is reset due to a new minimum validation loss. I would prefer the early stopping criteria be explicitly checked during the pre-train routine such that training never resumes if it was stopped due to early stopping, in the same way it never resumes if it was stopped due to the maximum number of epochs.

### Pitch

I was able to hack this together by overriding the EarlyStopping class and adding the following hook (using master, currently 1.4.0.dev0):

```python
    def on_pretrain_routine_end(
        self,
        trainer: 'pl.Trainer',
        pl_module: 'pl.LightningModule'
    ) -> None:
        # Check if we stopped previously
        if self.stopped_epoch > 0 and self.stopped_epoch <= trainer.current_epoch:
            trainer.should_stop = True
```

However I'm not sure if this is the most reliable way to do this, and I believe this should be automatically done. If the "state" of training was that it ended due to early stopping, when the system resumes the same state should be loaded, which caused training to stop and therefore should do the same thing again.

### Alternatives

I've considered running a full validation epoch to trigger this normally, but that seems unnecessary if the state already exists.

#### Thank you!

**Comment by tchaton** [comment]
Dear @collinmccarthy,

Yes, it is an interesting one.
So to sum up, if a checkpoint is generated after EarlyStopping stopped fitting and if you were to reload this checkpoint and call fit again, you would expect only validation to be run right and then stop ?

This makes sense to me.
Would you like to make a contribution ?

I think the cleanest would be save the training state in the checkpoint directly to make it available on reload.

Best,
T.C

**Comment by collinmccarthy** [comment]
@tchaton Sorry for the delayed response. 

> So to sum up, if a checkpoint is generated after EarlyStopping stopped fitting and if you were to reload this checkpoint and call fit again, you would expect only validation to be run right and then stop ?

I would expect neither training nor validation be run, at least if `trainer.fit()` were called again. In my hack I _think_ setting `trainer.should_stop = True` causes `trainer.fit()` to return immediately and not run any training nor validation epochs (after the validation sanity check).  

> Would you like to make a contribution ?

Yes, I can do that, but I'd like to spend a bit more time stepping through the PTL codebase first, and I have a deadline this month so it would have to be next month. If there's no immediate hurry from you guys then I'd be happy to give it a shot in the near future.

> I think the cleanest would be save the training state in the checkpoint directly to make it available on reload.

Do you mean saving the state in the checkpoint such that we can call `EarlyStopping._run_early_stopping_check()` during the pre-train routine? That's what I was thinking but I'm not super familiar with how to actually do that (yet). If that's not what you mean could you be more specific?

Thanks for taking a look at this!
-Collin
